abbyy_to_epub3/parse_abbyy.py

```diff
@@ -203,286 +269,239 @@ def parse_abbyy(self):
-                elif constants.OLD_NS in abbyy_nsm.values():
-                    self.nsm = constants.OLD_NSM
-                    self.ns = constants.OLD_NS
-                    self.version = "FR6"
-                else:
-                    raise RuntimeError("Input XML not in a supported schema.")
-                self.logger.debug("FineReader Version {}".format(self.version))
-                self.metadata['fr-version'] = self.version
-            else:
-                return
-
-    def decompose_xml(self, elem):
+    def process_styles(self, elem):
         """
-        iteratively parse the ABBYY file into data structures
-        so that intelligent parsing can happen later.
+        Iteratively parse styles from the ABBYY file into data structures.
         The ABBYY seems to be sometimes inconsistent about whether these
         elements have a namespace, so be forgiving.
         """
 
         if (
-            elem.tag == "{{{}}}fontStyle".format(self.ns) or
-            elem.tag == "fontStyle"
-        ):
-            self.fontStyles[elem.get("id")] = dict(elem.attrib)
-        elif (
             elem.tag == "{{{}}}paragraphStyle".format(self.ns) or
             elem.tag == "paragraphStyle"
         ):
             """
             Paragraph styles are on their own at the start of the ABBYY
-            and refer to sibling fontStyle elements
+            and contain child fontStyle elements
             """
-            id = elem.get("id")
-            attribs = dict(elem.attrib)
-            self.paragraphs[id] = attribs
-            if (
-                'mainFontStyleId' in attribs and
-                'mainFontStyleId' in self.fontStyles
-            ):
-                    self.paragraphs[id]['fontstyle'] = self.fontStyles[
-                        'mainFontStyleId'
-                    ]
+            self.paragraphs[elem.get("id")] = dict(elem.attrib)
+            fontstyles = elem.iterchildren()
+            for fontstyle in fontstyles:
+                self.fontStyles[fontstyle.get("id")] = dict(fontstyle.attrib)
+
+    def process_pages(self, elem):
+        """
+        Iteratively process pages from the ABBYY file. We have to process now
+        rather than copying the pages for later processing, because deepcopying
+        an lxml element replicates the entire tree.
+        The ABBYY seems to be sometimes inconsistent about whether these
+        elements have a namespace, so be forgiving.
+        """
 
-        elif (
+        if (
             elem.tag == "{{{}}}page".format(self.ns) or
             elem.tag == "page"
         ):
-            self.pages.append(deepcopy(elem))
-        else:
-            return
+            d = {'page_no': self.page_no}
 
-        # only garbage collect if we have found & processed a node
-        elem.clear()
-
-    def parse_content(self):
-        """ Parse each page of the book.  """
-        page_no = 0
-        d = {'page_no': page_no}
```

Comment:
Why are we setting `d` here if the only place its being used (line 331) it's being overwritten?

Reply:
Ah, the joy of refactoring.